File: apps/web/types/dataset.ts
export type DatasetStatus = "uploading" | "profiling" | "ready" | "failed";

export interface PreprocessChange {
  code: string;
  title: string;
  detail: string;
}

export interface DatasetTableSummary {
  name: string;
  filename: string;
  row_count: number | null;
  col_count: number | null;
}

export interface Dataset {
  id: string;
  filename: string;
  status: DatasetStatus;
  row_count: number | null;
  col_count: number | null;
  error: string | null;
  created_at: string;
  archived: boolean;
  preprocessed: boolean;
  // At upload: what cleaning WOULD do (drives the recommendation card).
  // After preprocessing: what was applied. null when the data is already clean.
  preprocessing_summary: PreprocessChange[] | null;
  // Per-table summary when the dataset spans multiple CSVs; null for single-table.
  tables: DatasetTableSummary[] | null;
}

export interface DatasetColumn {
  name: string;
  dtype: string;
  position: number;
  null_count: number;
  distinct_count: number;
  min_value: string | null;
  max_value: string | null;
  sample_values: unknown[];
}

export interface DatasetTableProfile extends DatasetTableSummary {
  columns: DatasetColumn[];
}

export interface DatasetProfile extends Dataset {
  columns: DatasetColumn[];
  // Full per-table profiles for multi-table datasets; null for single-table.
  tables: DatasetTableProfile[] | null;
}

File: apps/api/app/schemas/dataset.py
"""Dataset request/response DTOs."""
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.dataset import DatasetStatus


class DatasetUpdate(BaseModel):
    filename: str = Field(min_length=1, max_length=255)


class DatasetColumnRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    name: str
    dtype: str
    position: int
    null_count: int
    distinct_count: int
    min_value: str | None
    max_value: str | None
    sample_values: list


class PreprocessChange(BaseModel):
    code: str
    title: str
    detail: str


class DatasetTableSummary(BaseModel):
    """One table within a (possibly multi-CSV) dataset, without its columns."""

    model_config = ConfigDict(from_attributes=True)

    name: str
    filename: str
    row_count: int | None = None
    col_count: int | None = None


class DatasetTableProfile(DatasetTableSummary):
    columns: list[DatasetColumnRead] = []


class DatasetRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    filename: str
    status: DatasetStatus
    row_count: int | None
    col_count: int | None
    error: str | None
    created_at: datetime
    archived: bool = False
    preprocessed: bool = False
    # At ingest: what cleaning WOULD do (drives the recommendation card).
    # After preprocessing: what was applied.
    preprocessing_summary: list[PreprocessChange] | None = None
    # Per-table summary when the dataset spans multiple CSVs; None for single-table.
    tables: list[DatasetTableSummary] | None = None


class DatasetProfile(DatasetRead):
    """Full profile: dataset metadata plus the per-column summary."""

    columns: list[DatasetColumnRead]
    # Full per-table profiles (with columns) for multi-table datasets.
    tables: list[DatasetTableProfile] | None = None
